javaslang/src/main/java/javaslang/collection/HashArrayMappedTrie.java

```diff
@@ -51,6 +51,87 @@
         PUT, REMOVE
     }
 
+    class It<K, V> extends AbstractIterator<LeafNode<K, V>> {
+
+        private final static int MAX_LEVELS = Integer.SIZE / AbstractNode.SIZE + 1;
+
+        private final int total;
+        private final Object[] nodes = new Object[MAX_LEVELS];
+        private final int[] indexes = new int[MAX_LEVELS];
+
+        private int level;
+        private int ptr = 0;
+
+        It(AbstractNode<K, V> root) {
+            total = root.size();
+            level = downstairs(nodes, indexes, root, 0);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return ptr < total;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        protected LeafNode<K, V> getNext() {
+            Object node = nodes[level];
+            while (!(node instanceof LeafNode)) {
+                node = findNextLeaf();
+            }
+            ptr++;
+            if (node instanceof LeafList) {
+                LeafList<K, V> leaf = (LeafList<K, V>) node;
+                nodes[level] = leaf.tail;
+                return leaf;
+            } else {
+                nodes[level] = EmptyNode.instance();
+                return (LeafSingleton<K, V>) node;
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        private Object findNextLeaf() {
+            AbstractNode<K, V> node = null;
+            while (level > 0) {
+                level--;
+                indexes[level]++;
+                node = getChild((AbstractNode<K, V>) nodes[level], indexes[level]);
+                if (node != null) {
+                    break;
+                }
+            }
+            level = downstairs(nodes, indexes, node, level + 1);
+            return nodes[level];
+        }
+
+        private static <K, V> int downstairs(Object[] nodes, int[] indexes, AbstractNode<K, V> root, int level) {
+            while (true) {
+                nodes[level] = root;
+                indexes[level] = 0;
+                root = getChild(root, 0);
+                if (root == null) {
+                    break;
+                } else {
+                    level++;
+                }
+            }
+            return level;
+        }
+
+        @SuppressWarnings("unchecked")
+        private static <K, V> AbstractNode<K, V> getChild(AbstractNode<K, V> node, int index) {
+            if (node instanceof IndexedNode) {
+                Object[] subNodes = ((IndexedNode<K, V>) node).subNodes;
+                return index < subNodes.length ? (AbstractNode<K, V>) subNodes[index] : null;
+            } else if (node instanceof ArrayNode) {
+                ArrayNode<K, V> arrayNode = (ArrayNode<K, V>) node;
+                return index < 32 ? (AbstractNode<K, V>) arrayNode.subNodes[index] : null;
+            }
+            return null;
+        }
+    }
+
```

Comment:
Yep :smile: 
Last iterator was simple and obvious, but used a lot (O(n)) of temporary objects